### PR TITLE
Embed csv parser and dependencies directly into sparklyr

### PR DESCRIPTION
**Pull request by javierluraschi**
Several users have experienced problems installing the databricks' csv package from https://spark-packages.org/ due to local connectivity issues or security restrictions in closed-cluster environments.

This change embeds the csv package into `sparklyr` and it's cone dependency.

### CODE REVIEW

**Comment by jjallaire** [comment]
Interesting juxtaposition between this PR and this issue: https://github.com/rstudio/sparklyr/pull/314

One saying "please make everything a Spark package for ease of deployment" and the other saying "please make nothing a Spark package for ease of deployment".

I wonder if there is a way for us to support both approaches via a global option.

File: R/spark_shell.R [hunk 1]
```diff
@@ -1,4 +1,11 @@
+shell_connection_validate_config <- function(config) {
+  if ("spark.jars.default" %in% config) {
```
Comment (hadley) [inline R/spark_shell.R]: Shouldn't that be `names(config)`?

Comment (javierluraschi) [inline R/spark_shell.R]: Yes!

**Comment by javierluraschi** [comment]
@jjallaire currently, if someone doesn't want the csv package, they could remove this form the `config.yml` file, I wonder if that's enough as a global option.

Is https://github.com/rstudio/sparklyr/issues/315 the issue you are referring to? One interesting thing to take a look at, is as to why the R session crashes when the java process crashes. Haven't taken a look at this yet but maybe this already sounds familiar? We start the java process through a bash script running through `system2`, I would expect that if this crashes, it has no effect on the R session, but perhaps this is caused by the socket connection terminating unexpectedly or something else.

**Comment by javierluraschi** [comment]
Ah, this issue https://github.com/rstudio/sparklyr/issues/171! Yes, that's interesting since the argument there would be to pull everything out of `sparklyr`, as you mention, we probably need both paths.

**Comment by falaki** [comment]
Note that CSV data source has been inlined in Spark 2.x. Including csv Jar file with spark 2.x will cause class loader problems.

File: inst/conf/config-template.yml [hunk 2]
```diff
@@ -4,9 +4,13 @@ default:
   spark.sql.shuffle.partitions.local: !expr parallel::detectCores()
   spark.env.SPARK_LOCAL_IP.local: 127.0.0.1
 
+  # default embedded packages to load
+  sparklyr.jars.default:
+    - !expr system.file(file.path("java", "spark-csv_2.11-1.3.0.jar"), package = "sparklyr")
+    - !expr system.file(file.path("java", "commons-csv-1.1.jar"), package = "sparklyr")
```
Comment (falaki) [inline inst/conf/config-template.yml]: You may also want to include uniVocity-parsers jar file. spark-csv supports an option to chose parser library. Some features only work with univocity. Also there is slight difference in parsing behavior.

Comment (javierluraschi) [inline inst/conf/config-template.yml]: Good point, added.

**Comment by javierluraschi** [comment]
Thanks @falaki, I was not aware that the `spark-csv` package is now included in Spark 2.0. `sparklyr` will default to use the inline version from Spark 2.0 onwards, otherwise, the embedded version will be deployed by default.

**Comment by javierluraschi** [comment]
Planing to merge this soon.

**Pull request merged**